[refactor] OpenApiApartmentPriceService 리팩토링

rlqor9688 opened the pull request:
## 📌 관련 이슈
- Closes #93 93

## ✨ 변경 사항
[Service]
1. OpenApiApartmentPriceService : API 호출 1번에 Redis 템플릿에 저장하는 메서드 구현 / 최신 데이터 받아오는 메서드 리팩토링
2. CachedApartmentPriceService : 평균 상승률 계산 시 입력으로 받지 않고, 고정된 값으로 반환하는 방식으로 리팩토링

[Scheduler]
1. OpenApiApartmentPriceService의 fetchDurationAll(int year) 메서드를 활용해 데이터를 업데이트 하는 방식으로 변환. 모든 데이터를 저장하는 것이 아닌, 가장 최신, 가장 오래된 데이터만 저장

[Repository]
1. ApartmentPriceCacheRepository : findLatest 메서드 추가, findStartAndEnd 메서드 입력 필드 중 연도 입력 삭제
2. RedisApartmentPriceRepository : 구조 리팩토링. Redis에서 값 읽기/ Redis로 값 쓰기 메서드 구현. 데이터 중 가장 오래된 값, 가장 최신 값을 함께 저장하는 메서드 구현

[Config]
1. RestTemplateConfig : timeout 설정을 위한 SimpleClientHttpRequestFactory 설정 추가

[Common]
1. ApiResponse : 성공 응답 시 ErrorCode 반환 안 하도록 수정

## 📷 Postman 
- OpenApiApartmentPriceServiceTest
![image](https://github.com/user-attachments/assets/057f5cbd-7e11-4a65-bad2-a4e8a632074c)

- CachedApartmentPriceServiceTest
![image](https://github.com/user-attachments/assets/3fe097dc-0098-40e9-b4ab-5f496763f4fc)

## ✅ 체크리스트
- [ ] 관련 이슈에 연결됨
- [ ] 기능이 정상적으로 동작함
- [ ] 코드 리뷰를 받았음
- [ ] 불필요한 로그/코드 제거됨

## 💬 기타 참고 사항
- 테스트 방법, 추가 설명 등 필요한 경우 작성해 주세요.

rlqor9688 committed:
[refactor] OpenApiApartmentPriceService 리팩토링
[Service]
1. OpenApiApartmentPriceService : API 호출 1번에 Redis 템플릿에 저장하는 메서드 구현 / 최신 데이터 받아오는 메서드 리팩토링
2. CachedApartmentPriceService : 평균 상승률 계산 시 입력으로 받지 않고, 고정된 값으로 반환하는 방식으로 리팩토링

[Scheduler]
1. OpenApiApartmentPriceService의 fetchDurationAll(int year) 메서드를 활용해 데이터를 업데이트 하는 방식으로 변환. 모든 데이터를 저장하는 것이 아닌, 가장 최신, 가장 오래된 데이터만 저장

[Repository]
1. ApartmentPriceCacheRepository : findLatest 메서드 추가, findStartAndEnd 메서드 입력 필드 중 연도 입력 삭제
2. RedisApartmentPriceRepository : 구조 리팩토링. Redis에서 값 읽기/ Redis로 값 쓰기 메서드 구현. 데이터 중 가장 오래된 값, 가장 최신 값을 함께 저장하는 메서드 구현

[Config]
1. RestTemplateConfig : timeout 설정을 위한 SimpleClientHttpRequestFactory 설정 추가

[Common]
1. ApiResponse : 성공 응답 시 ErrorCode 반환 안 하도록 수정

byeongtaek12 commented on src/main/java/org/example/lifechart/domain/goal/scheduler/ApartmentPriceSyncScheduler.java:
```diff
@@ -24,18 +23,22 @@ public class ApartmentPriceSyncScheduler {
+
+				redisRepository.saveWithAlias(start, true); // key: apt-price:서울:동남권:start
+				redisRepository.saveWithAlias(end, false); // key: apt-price:서울:동남권:end
 			}
```
왜 같은 값을 다른 키에 저장하지라고 생각했는데 gpt한테 물어보니 날짜를 몰라도 시작 값과 끝 값을 쉽게 조회할 수 있다고 해서 바로 이해했습니다.

byeongtaek12 commented on src/main/java/org/example/lifechart/domain/goal/service/CachedApartmentPriceService.java:
```diff
@@ -27,7 +27,7 @@ public Long getAveragePrice(String region, String subregion, Long area) {
 	}
 
-	@Override // 미래 시점의 가격 계싼
+	@Override // 미래 시점의 가격 계산
```
아 일부로 말 안해드렸는데 찾으셨네요 계싼 웃겼는데

rlqor9688 replied:
ㅋㅋㅋㅋ 일부로 말 안하셨다니...

KSP0321 replied:
ㅋㅋㅋㅋㅋㅋ

byeongtaek12 commented on src/main/java/org/example/lifechart/domain/goal/service/OpenApiApartmentPriceService.java:
```diff
@@ -53,38 +57,76 @@ public ApartmentPriceDto fetchLatest(String region, String subregion) {
-			Map<String, Object> entry = latestEntry.get();
+				.max(Comparator.comparing(item -> (String) item.get("PRD_DE")))
+				.orElseThrow(() -> new CustomException(ErrorCode.DATA_NOT_FOUND));
 
```
👍

byeongtaek12 commented on src/main/java/org/example/lifechart/domain/goal/service/OpenApiApartmentPriceService.java:
```diff
@@ -53,38 +57,76 @@ public ApartmentPriceDto fetchLatest(String region, String subregion) {
-				.build();
+			return mapToDto(latestEntry, regionCode);
 		} catch (CustomException e) {
 			throw e;
```
어떤 예외가 발생했는지 로그를 찍으면 좋을 것 같습니다. 그리고 밖으로 예외를 던지고 있는데 외부에서 이 예외를 처리해주는 곳이 따로 있나요?

rlqor9688 replied:
로직 안에서 발생한 CustomException (DATA_NOT_FOUND, INVALID_REGION_MATCH)를 던져줘서 throw new CustomException(ErrorCode.~~) 처럼 동작한다고 생각했습니다! GPT 한테 물어보니 이렇게 던지면 GlobalException에서 잡아서 처리해주는데, 로그를 남기면 더 좋다고 하네요.! 로그 추가하겠습니다!
![image](https://github.com/user-attachments/assets/4e8f35fd-f8bc-45b6-b44c-6aea934f3737)

byeongtaek12 replied:
오 글로벌에서 처리해주는군요

rlqor9688 replied:
맞습니다! CustomException을 캐치해서 던지는 거라 글로벌 익셉션에서 잡아줍니다!

byeongtaek12 commented on src/main/java/org/example/lifechart/domain/goal/service/OpenApiApartmentPriceService.java:
```diff
@@ -53,38 +57,76 @@ public ApartmentPriceDto fetchLatest(String region, String subregion) {
+
+					return Map.entry(code,
+						Pair.of(mapToDto(oldest, code), mapToDto(latest, code)));
+				})
```
여기 oldest와 latest 둘다 null이 나올 수 있는 것 같은데 mapToDto에 null, code 이런 식으로 들어가면 npe 발생할 것 같습니다.

rlqor9688 replied:
오 감사합니다! 수정해서 커밋하겠습니다!

byeongtaek12 reviewed (approved):
코멘트 남긴 부분들 한번 확인 해주시고 전체 데이터 가져오는 방식보다 시작과 끝만 가져와서 저장하는 방식으로 리팩토링하신 부분 좋은 것 같습니다. 👍

rlqor9688 committed:
[fix] NPE 방지 코드, Goal-Like, Comment 요청 코드 추가
[Repository]
1. CommentRepository : deleteAllByGoalId 메서드 추가
2. LikeRepository : deleteAllByGoalId 메서드 추가

[Service]
1. OpenApiApartmentPriceService: 에러 로그 추가
2. GoalServiceImpl : CommentRepository, LikeRepository 필드 추가

[Entity]
1. clearComment, clearLike 메서드 추가

[GoalServiceImplTest] : comment, like 삭제 테스트 로직 추가

byeongtaek12 commented:
새로 커밋하신 것도 확인 완료했습니다. 수고많으셨습니다!

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-openapi-develop

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-openapi-develop

rlqor9688 committed:
[feat] GoalRepository 요청 코드 추가
Simulation 파트 요청 코드 추가

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-openapi-develop

woneyee commented on src/main/java/org/example/lifechart/domain/goal/repository/RedisApartmentPriceRepository.java:
```diff
@@ -34,104 +24,73 @@ public class RedisApartmentPriceRepository implements ApartmentPriceCacheReposit
-		return periods.stream()
-			.max(Comparator.naturalOrder()) // 문자열 정렬
-			.orElseThrow(() -> new CustomException(ErrorCode.DATA_NOT_FOUND));
+		return Optional.empty();
```
Redis에 데이터가 없거나 에러 발생시 optional.empty보다 재조회하는 방법도 고려할 수 있을 것 같습니다 보완방법이니 참고차 알려드립니다 !

woneyee reviewed (approved):
레디스라서 ..어렵네요 확인끝났습니다

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-openapi-develop
